src/Processors/Formats/Impl/DiagramOutputFrormat.cpp

```diff
@@ -0,0 +1,649 @@
+    "\u28D0", "\u28D1", "\u28D2", "\u28D3", "\u28D4", "\u28D5", "\u28D6", "\u28D7", "\u28D8", "\u28D9", "\u28DA", "\u28DB", "\u28DC",
+    "\u28DD", "\u28DE", "\u28DF", "\u28E0", "\u28E1", "\u28E2", "\u28E3", "\u28E4", "\u28E5", "\u28E6", "\u28E7", "\u28E8", "\u28E9",
+    "\u28EA", "\u28EB", "\u28EC", "\u28ED", "\u28EE", "\u28EF", "\u28F0", "\u28F1", "\u28F2", "\u28F3", "\u28F4", "\u28F5", "\u28F6",
+    "\u28F7", "\u28F8", "\u28F9", "\u28FA", "\u28FB", "\u28FC", "\u28FD", "\u28FE", "\u28FF",
+};
+
+DiagramSymbols utf_diagram_symbols;
+
+DiagramSymbols ascii_diagram_symbols{"+", "+", "+", "+", "+", "+", "+", "+", "+", "+", "+",
+                                     "+", "+", "+", "-", "-", "|", "|", "#", "~", "^"};
+}
+
+
+void DiagramOutputFormat::write()
+{
+    if (!diagram_type.contains(format_settings.diagram.diagram_type))
+        throw Exception(ErrorCodes::UNKNOWN_SETTING, "No such diagram type.");
+    switch (diagram_type.at(format_settings.diagram.diagram_type))
+    {
+        case DiagramType::SCATTER:
+            writeScatter();
+            break;
+        case DiagramType::LINE:
+            writeLineplot();
+            break;
+        case DiagramType::HISTOGRAM:
+            writeHistogram();
+            break;
+        case DiagramType::HEAT_MAP:
+            writeHeatMap();
+            break;
+    }
+}
+
+DiagramOutputFormat::DiagramOutputFormat(WriteBuffer & out_, const Block & header_, const FormatSettings & format_settings_)
+    : IOutputFormat(header_, out_), format_settings{format_settings_}, serializations(header_.getSerializations())
+{
+    is_ascii_symbols = format_settings_.diagram.is_ascii_symbols;
+    height = DiagramConstants::DEFAULT_HEIGHT;
+    width = DiagramConstants::DEFAULT_WIEDTH;
+}
+
+void DiagramOutputFormat::consume(Chunk chunk)
+{
+    chunks.push_back(std::move(chunk));
+}
+
+
+void DiagramOutputFormat::writeScatter()
+{
+    const DiagramSymbols & diagram_symbols = is_ascii_symbols ? ascii_diagram_symbols : utf_diagram_symbols;
+    if (format_settings.diagram.limit_height > 0)
+        height = format_settings.diagram.limit_height;
+    if (format_settings.diagram.limit_width > 0)
+        width = format_settings.diagram.limit_width;
+    std::vector<std::vector<BraileSymbol>> plot = std::vector<std::vector<BraileSymbol>>(height, std::vector<BraileSymbol>(width));
+
+    // Calculate data limits for scaling
+    height *= DiagramConstants::BRAILE_BLOCK_HEIGHT;
+    width *= DiagramConstants::BRAILE_BLOCK_WIDTH;
+    Float64 max_x = std::numeric_limits<Float64>::min();
+    Float64 max_y = std::numeric_limits<Float64>::min();
+    Float64 min_x = std::numeric_limits<Float64>::max();
+    Float64 min_y = std::numeric_limits<Float64>::max();
+    for (const auto & chunk : chunks)
+    {
+        if (chunk.getNumColumns() != DiagramConstants::EXPEXTED_COLUMN_NUMBER)
+            throw Exception(ErrorCodes::INCORRECT_NUMBER_OF_COLUMNS, "Incorrect number of columns");
```

Comment:
It will be better to check number of columns and data types in `DiagramOutputFormat` constructor using provided `header`, otherwise this exception will be thrown only when all the data is collected in memory.